File: Apps/iOS8/AlarmClock/AlarmClock/ViewController.m
//
//  ViewController.m
//  AlarmClock
//
//  Created by Clemens Wagner on 17.07.13.
//  Copyright (c) 2013 Cocoaneheads. All rights reserved.
//

#import "ViewController.h"
#import "ClockView.h"
#import "ClockControl.h"
#import "UIView+AlarmClock.h"

const NSTimeInterval kSecondsOfDay = 60.0 * 60.0 * 24.0;

@interface ViewController ()

@property (weak, nonatomic) IBOutlet ClockView *clockView;
@property (weak, nonatomic) IBOutlet ClockControl *clockControl;
@property (weak, nonatomic) IBOutlet UILabel *timeLabel;

@property (nonatomic) BOOL alarmHidden;

- (IBAction)updateTimeLabel;
- (IBAction)switchAlarm:(UILongPressGestureRecognizer *)inRecognizer;
- (IBAction)updateAlarm;

@end

@implementation ViewController

-(BOOL) prefersStatusBarHidden {
    return YES;
}

- (void)viewDidLoad {
    [super viewDidLoad];
	// Do any additional setup after loading the view, typically from a nib.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewWillAppear:(BOOL)inAnimated {
    [super viewWillAppear:inAnimated];
    [self updateViews];
}

- (void)viewDidAppear:(BOOL)inAnimated {
    [super viewDidAppear:inAnimated];
    [self.clockView startAnimation];
}

- (void)viewWillDisappear:(BOOL)inAnimated {
    [self.clockView stopAnimation];
    [super viewWillDisappear:inAnimated];
}

- (void)updateViews {
    UIApplication *theApplication = [UIApplication sharedApplication];
    UILocalNotification *theNotification = [[theApplication scheduledLocalNotifications] lastObject];

    if(theNotification == nil) {
        self.alarmHidden = YES;
    }
    else {
        NSTimeInterval theTime = [theNotification.fireDate timeIntervalSinceReferenceDate] - self.startTimeOfCurrentDay;

        theTime = remainder(theTime, kSecondsOfDay / 2.0);
        self.clockControl.time = theTime < 0 ? theTime + kSecondsOfDay / 2.0 : theTime;
        self.alarmHidden = NO;
    }
    [self updateTimeLabel];
}

- (BOOL)alarmHidden {
    return self.clockControl.hidden;
}

- (void)setAlarmHidden:(BOOL)inAlarmHidden {
    self.clockControl.hidden = inAlarmHidden;
    self.timeLabel.hidden = inAlarmHidden;
}

- (IBAction)updateTimeLabel {
    NSInteger theTime = round(self.clockControl.time / 60.0);
    NSInteger theMinutes = theTime % 60;
    NSInteger theHours = theTime / 60;

    self.timeLabel.text = [NSString stringWithFormat:@"%d:%02d",
                           (int)theHours, (int)theMinutes];
}

- (IBAction)switchAlarm:(UILongPressGestureRecognizer *)inRecognizer {
    if(inRecognizer.state == UIGestureRecognizerStateEnded) {
        if(self.alarmHidden) {
            CGPoint thePoint = [inRecognizer locationInView:self.clockView];
            CGFloat theAngle = [self.clockView angleWithPoint:thePoint];
            NSTimeInterval theTime = 21600.0 * theAngle / M_PI;

            self.alarmHidden = NO;
            self.clockControl.time = theTime;
            [self updateTimeLabel];
        }
        else {
            self.alarmHidden = YES;
        }
        [self updateAlarm];
    }
}

- (IBAction)updateAlarm {
    if(self.alarmHidden) {
        UIApplication *theApplication = [UIApplication sharedApplication];

        [theApplication cancelAllLocalNotifications];
    }
    else {
        [self createAlarm];
    }
}


- (NSDate *)alarmDate {
    NSTimeInterval theTime = self.startTimeOfCurrentDay + self.clockControl.time;
    
    while(theTime < [NSDate timeIntervalSinceReferenceDate]) {
        theTime += kSecondsOfDay / 2.0;
    }
    return [NSDate dateWithTimeIntervalSinceReferenceDate:theTime];
}

- (void)createAlarm {
    UIApplication *theApplication = [UIApplication sharedApplication];
    UILocalNotification *theNotification = [UILocalNotification new];

    [theApplication cancelAllLocalNotifications];
    theNotification.fireDate = [self alarmDate];
    theNotification.timeZone = [NSTimeZone defaultTimeZone];
    theNotification.alertBody = NSLocalizedString(@"Wake up", @"Alarm message");
    theNotification.soundName = UILocalNotificationDefaultSoundName;
    [theApplication scheduleLocalNotification:theNotification];
}

- (NSTimeInterval)startTimeOfCurrentDay {
    NSCalendar *theCalendar = [NSCalendar currentCalendar];
    NSDateComponents *theComponents = [theCalendar components:NSCalendarUnitYear | NSCalendarUnitMonth |NSCalendarUnitDay fromDate:[NSDate date]];
    NSDate *theDate = [theCalendar dateFromComponents:theComponents];

    return [theDate timeIntervalSinceReferenceDate];
}

- (NSString *)description {
    return [NSString stringWithFormat:@"alarm: %@ (%@)", self.timeLabel.text, self.alarmHidden ? @"off" : @"on"];
}

- (NSString *)debugDescription {
    return [NSString stringWithFormat:@"debug alarm: %@ (%.3fs, %@)",
            self.timeLabel.text, self.clockControl.time,
            self.alarmHidden ? @"off" : @"on"];
}

@end

File: Apps/iOS8/Swift/AlarmClock/AlarmClock/ViewController.swift
//
//  ViewController.swift
//  AlarmClock
//
//  Created by Clemens Wagner on 21.06.14.
//  Copyright (c) 2014 Clemens Wagner. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    let kSecondsOfDay:NSTimeInterval = 60.0 * 60.0 * 24.0
    
    @IBOutlet weak var clockView: ClockView!
    @IBOutlet weak var clockControl: ClockControl!
    @IBOutlet weak var timeLabel: UILabel!
    
    var alarmHidden: Bool {
    get {
        return clockControl.hidden
    }
    set {
        clockControl.hidden = newValue
        timeLabel.hidden = newValue
    }
    }
    
    override func prefersStatusBarHidden() -> Bool {
        return true
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewWillAppear(inAnimated: Bool) {
        super.viewWillAppear(inAnimated)
        updateViews()
    }
    
    override func viewDidAppear(inAnimated: Bool) {
        super.viewDidAppear(inAnimated)
        clockView.startAnimation()
    }
    
    override func viewWillDisappear(inAnimated: Bool) {
        clockView.stopAnimation()
        super.viewWillDisappear(inAnimated)
    }
    
    func updateViews() {
        let theApplication = UIApplication.sharedApplication()
        let theNotifications = theApplication.scheduledLocalNotifications
        let theNotification : UILocalNotification? = theNotifications.last as? UILocalNotification
        
        if(theNotification == nil) {
            alarmHidden = true;
        }
        else {
            var theTime = theNotification!.fireDate!.timeIntervalSinceReferenceDate - startTimeOfCurrentDay()
            
            theTime = theTime % (kSecondsOfDay / 2.0)
            clockControl.time = theTime < 0 ? theTime + kSecondsOfDay / 2.0 : theTime;
            alarmHidden = false
        }
        updateTimeLabel()
    }
    
    @IBAction func updateTimeLabel() {
        let theTime = UInt(round(clockControl.time / 60.0))
        let theMinutes = theTime % 60
        let theHours = theTime / 60
        
        timeLabel.text = NSString(format:"%d:%02d", theHours, theMinutes)
    }
    
    @IBAction func switchAlarm(inRecognizer:UILongPressGestureRecognizer!) {
        if(inRecognizer.state == UIGestureRecognizerState.Ended) {
            if(alarmHidden) {
                let thePoint = inRecognizer.locationInView(clockView)
                let theAngle = Double(clockView.angleWithPoint(thePoint))
                let theTime = 21600.0 * theAngle / M_PI
                
                alarmHidden = false
                clockControl.time = theTime
                updateTimeLabel()
            }
            else {
                alarmHidden = true
            }
            updateAlarm()
        }
    }
    
    @IBAction func updateAlarm() {
        if(alarmHidden) {
            let theApplication = UIApplication.sharedApplication()
            
            theApplication.cancelAllLocalNotifications()
        }
        else {
            createAlarm()
        }
    }
    
    func alarmDate() -> NSDate {
        var theTime:NSTimeInterval = startTimeOfCurrentDay() + clockControl.time;
        
        while(theTime < NSDate.timeIntervalSinceReferenceDate()) {
            theTime += kSecondsOfDay / 2.0;
        }
        return NSDate(timeIntervalSinceReferenceDate:theTime)
    }
    
    func createAlarm() {
        let theApplication = UIApplication.sharedApplication()
        let theNotification = UILocalNotification()
        
        theApplication.cancelAllLocalNotifications()
        theNotification.fireDate = alarmDate()
        theNotification.timeZone = NSTimeZone.defaultTimeZone()
        theNotification.alertBody = NSLocalizedString("Wake up", comment:"Alarm message")
        theNotification.soundName = "ringtone.caf"
        theApplication.scheduleLocalNotification(theNotification)
    }
    
    func startTimeOfCurrentDay() -> NSTimeInterval {
        let theCalendar = NSCalendar.currentCalendar()
        let theComponents = theCalendar.components(NSCalendarUnit.YearCalendarUnit | NSCalendarUnit.MonthCalendarUnit | NSCalendarUnit.DayCalendarUnit, fromDate: NSDate())
        let theDate = theCalendar.dateFromComponents(theComponents)
        
        return theDate!.timeIntervalSinceReferenceDate
    }
    
    func description() -> NSString {
        return NSString(format:"alarm: %@ (%@)", timeLabel.text!, alarmHidden ? "off" : "on")
    }
    
    func debugDescription() -> NSString {
        return NSString(format:"debug alarm: %@ (%.3fs, %@)", timeLabel.text!, clockControl.time, alarmHidden ? "off" : "on")
    }
}

